### DISCUSSION TITLE
http server in tauri app

### DISCUSSION THREAD

**Discussion by aknuth** [tauri-apps/tauri#2751] [Q&A] [answered]
Hi all,
I want to let my tauri app receive http REST requests (GET, PUT; POST). How can I implement this ? I can't see a JS API (even not on the examples sites - https://github.com/tauri-apps/tauri/tree/next/examples) which solves this ...
Any ideas?

**Comment by FabianLars** [comment, marked as answer]
You have to bundle a server yourself, either in rust or via a sidecar ( [docs](https://tauri.studio/en/docs/usage/guides/bundler/sidecar) / [example](https://github.com/tauri-apps/tauri/tree/next/examples/sidecar) ).
Or if you already know a crazy way to host a server inside of a browser you probably can make that work too 🤷 (at least if i understood your usecase correctly)

**Reply by aknuth** [reply]
ok thanks got it and for sure no way to host a server inside a browser 😉 ... 
means either setup a node.js server and bundle it via sidecar or use e.g. rocket.rs as a rust REST server.

**Reply by conanchen** [reply]
I embedded actix-web.rs web server in tauri-app. 

1. tried to bundle my app to run and found [security policy](https://github.com/tauri-apps/tauri/issues/2002)
2. created ca for the embedded server using mkcert to enable https.  
3. the app works fine on the same pc where ca created by mkcert command.
4. If try to install the app into another pc , get the same  [security policy](https://github.com/tauri-apps/tauri/issues/2002)
5. then have to recreate cert for the pc to be installed, 
6. and for many other pcs, that I have to create certs, is it feasible  since there is no cert for localhost?

is there other way?

**Reply by Ripwords** [reply]
> ok thanks got it and for sure no way to host a server inside a browser 😉 ... means either setup a node.js server and bundle it via sidecar or use e.g. rocket.rs as a rust REST server.

I also ended up going with rocket.rs however because I am new to rust, I'm struggling to get both rocket.rs and tauri to run at the same time. Threads didn't work for me ("thread '<unnamed>' panicked at 'Initializing the event loop outside of the main thread is a significant cross-platform compatibility hazard") and maybe it's because I'm not doing it the right way. 

I would appreciate some pointers if you got it working.

**Reply by FabianLars** [reply]
@Ripwords That error is caused by #3172.
You need to run tauri on the main thread and rocket in a secondary one. The most basic setup would look like this:
```rs
#[rocket::get("/hello")]
fn world() -> &'static str {
  "Hello, world!"
}

fn main() {
  tauri::Builder::default()
    .setup(|app| {
      tauri::async_runtime::spawn(
        rocket::build().mount("/hello", rocket::routes![world]).launch()
      );
      Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

**Reply by Ripwords** [reply]
I see, the tauri::async_runtime was what I was trying to find, thanks!

**Reply by eacevedof** [reply]
Hello guys. Did you solve this issue?

**Reply by richardeschloss** [reply]
Fwiw, I found rocket to be the easiest solution. As somewhat of a Tauri newbie, I found it far easier than the sidecar example. I think others may find it useful if the rocket snippet were highlighted at the top in the "accepted answer".

**Reply by tom-haines** [reply]
As a postnote to this thread, I found it simpler to use Actix Web due to issues with Rocket asking for rust nightly features (under v0.4 at least).

```rust

#[get("/")]
async fn hello() -> impl Responder {
    HttpResponse::Ok().body("Hello world!")
}

fn main() {
  tauri::Builder::default()
    .setup(|app| {
        tauri::async_runtime::spawn(
          HttpServer::new(|| {
              App::new()
                  .service(hello)
          })
              .bind(("127.0.0.1", 8080))?
              .run()
        );
        Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}

```

**Reply by simonhyll** [reply]
Keep in mind that while you CAN make a server like this there's a fair number of security related aspects to be aware with this sort of approach. For example, you have to keep in mind that unless you yourself take some steps to prevent it, your API will be accessible to any other program on the users computer as well. And, if you select a port statically, what if that port is taken? You'll need to set it up to handle dynamic port assignment. And why are you really making the server? In most cases what you're getting is essentially the same thing as a Tauri command, just worse.

In short, I don't recommend starting a server in your Tauri app, unless you have a good grasp on related potential issues and have a plan for tackling them.

**Reply by tom-haines** [reply]
Agree @simonhyll as to being cautious, and that dynamic port allocation is a better tact to take.  

There are some valid use cases for a local http api, such as an oAuth URL handler to securely convert/store a transient credential for example.  Many apps (Google Drive etc) adopt this approach (a custom protocol handler is cleaner still as it removes the need for opening a port altogether, but I couldn't find any concrete examples of how to setup a url handler under tauri).  Most IDP providers I've seen allow a wildcard for the port (wildcard permitted, for localhost only, using MS/Google IDPs for example).  

For brevity, the example above is simply a hello world example.

**Reply by clytras** [reply]
@simonhyll one use cause for creating a server, is to create an HTTP proxy and bypass CORS restrictions inside Web Workers where neither IPC or Tauri HTTP client are available. See https://github.com/tauri-apps/tauri/issues/3308.

**Comment by jovanlopez32** [comment]
Hello how I implement Rocket inside tauri

**Comment by bingryan** [comment]
I also have a similar QA: https://github.com/tauri-apps/tauri/discussions/6731

**Reply by tom-haines** [reply]
I was able to access via http in production mode on mac, using sample code above.

**Comment by snilan** [comment]
Hi I know this is an old thread, but I am able to run a web server through Actix Web but I want to trigger an event to the frontend when I get that request. Is this possible?

I am new to Rust and can't figure out how to get a proper reference to the app handle.

```rs
let app_handler = app.handle();
tauri::async_runtime::spawn(
          HttpServer::new(move || {
              App::new().route(
                     "/",
                     web::get().to(|| async {
                            app_handler.emit(
                                "hello",
                                Some(Payload {
                                    message: "Hello from Rust!".to_string(),
                                }),
                            );
                            HttpResponse::Ok().body("hello world!")
                        }),
                    )
                })
                .bind(("127.0.0.1", 8080))?
                .run(),
);
```

**Reply by FabianLars** [reply]
I would probably store it in a state: https://actix.rs/docs/application/#state 🤔

**Reply by mandaputtra** [reply]
did you managed to do it? @snilan. Im also having problem making it a global variables

```rs
// Prevents additional console window on Windows in release, DO NOT REMOVE!!
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
use actix_web::{get, App, HttpResponse, HttpServer, Responder};
use std::thread_local;
use tauri::{App as TauriApp, Manager};

thread_local! {
    static APP: Option<TauriApp> = None;
}

#[get("/")]
async fn hello() -> impl Responder {
    HttpResponse::Ok().body("Hello world!")
}

#[get("/actions?<action>")]
async fn handle_actions(action: String) -> impl Responder {
    APP.with(|app| {
        if let Some(app) = app.as_ref() {
            match action.as_str() {
                "session_end" => {
                    app.get_window("main").unwrap().show().unwrap();
                    HttpResponse::Ok().body("Session ending, app shown")
                }
                "session_start" => {
                    app.get_window("main").unwrap().hide().unwrap();
                    HttpResponse::Ok().body("Session starting, app hidden")
                }
                _ => HttpResponse::Ok().body("Unknown action"),
            }
        } else {
            HttpResponse::InternalServerError().body("App instance not available")
        }
    })
}

fn main() {
    let app = tauri::Builder::default()
        .setup(|app| {
            *APP.with(|app| *app = Some(app));
            Ok(())
        })
        .run(tauri::generate_context!())
        .expect("error while running Tauri application");

    tauri::async_runtime::spawn(
      HttpServer::new(|| App::new()
        .service(hello)
        .service(handle_actions))
        .bind(("127.0.0.1", 8080))?
        .run()
    );
}
```

It alwasys got error

```
mismatched types
expected struct `tauri::App`
found reference `&Option<tauri::App>`
```

**Comment by ImageDeeply** [comment]
@snilan @mandaputtra - any update? Are you using Tauri 1.x or 2.x beta?

Or, are there better ways to enable customers to send commands to my app and get back a response?

**Reply by mandaputtra** [reply]
https://stackoverflow.com/a/77936504/5672961

Here I solve it using this

**Comment by yuyang-ok** [comment]
I have been wonder this too

**Reply by mandaputtra** [reply]
Here use this as an example https://stackoverflow.com/a/77936504/5672961

**Comment by lcdss** [comment]
Creating a HTTP Server in a Tauri app is useful to share functionality with smartphones in your local network. Or is there any other way to call a Tauri command on a phone when the Tauri app is running on a computer in the same network?

**Comment by lcdss** [comment]
I ended up doing something like this (in case anyone wants to do the same) to start a web server and share the Tauri App Handle with Actix-web requests:

```rs
tauri::Builder::default()
        .setup(|app| {
            // Only start the server if we are in a desktop environment
            #[cfg(desktop)]
            {
                let app_handle = app.handle().clone();

                thread::spawn(|| {
                    server::init(app_handle).unwrap();
                });
            }

            Ok(())
        })
```

```rs
struct AppState {
    tauri: AppHandle,
}

#[actix_web::main]
pub async fn init(tauri: AppHandle) -> io::Result<()> {
    let app_data = web::Data::new(AppState { tauri });

    let http_server = HttpServer::new(move || {
        let cors = Cors::default()
            .allow_any_origin()
            .allow_any_method()
            .allow_any_header()
            .max_age(3600);

        App::new()
            .app_data(app_data.clone())
            .wrap(cors)
            .wrap(middleware::Logger::default())
            .service(handlers::system::state)
    })
    .bind(("0.0.0.0", 1111))
    .expect("failed to bind address")
    .run();

    println!("Server started at http://0.0.0.0:1111");

    http_server.await
}
```

```rs
#[post("/system/{action}")]
async fn state(
    (action, payload, data): (Path<String>, Json<StatePayload>, Data<AppState>),
) -> HttpResponse {
    let shell = data.tauri.shell();
    ...
}
```
